aten/src/ATen/mps/MPSProfiler.mm

```diff
@@ -0,0 +1,306 @@
+  }
+}
+
+void MPSProfiler::beginProfileKernel(const void* handle, const std::string& kernelName, const TensorList& tensors) {
+  if (isKernelProfilingEnabled()) {
+    std::string profilerStrKey = Profiler::KernelInfo::buildKernelString(kernelName, tensors);
+    beginProfileKernel(handle, profilerStrKey, false);
+  }
+}
+
+void MPSProfiler::endProfileKernel(const void* handle) {
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(m_kernel_info_list.count(handle), "Failed to get kernel information!");
+  auto& kernelInfo = m_kernel_info_list[handle];
+
+  SignpostTypes signpostType = kernelInfo->type == Profiler::BaseInfo::Type::GRAPH ?
+                               SignpostTypes::RUN_MPSGRAPH : SignpostTypes::RUN_KERNEL;
+  addProfilerCompletedHandler(kernelInfo.get(), signpostType, SyncType::COMMIT);
+}
+
+void MPSProfiler::beginProfileCopy(const void* srcBuffer, const void* dstBuffer,
+                                   const OptionalTensorRef srcTensor,
+                                   const OptionalTensorRef dstTensor,
+                                   size_t length, bool isNonBlocking) {
+  if (!isCopyProfilingEnabled()) {
+    return;
+  }
+  auto copyInfo = std::make_unique<Profiler::CopyInfo>(dstBuffer, length, ++m_copy_counter, isNonBlocking);
+  copyInfo->srcStrKey = Profiler::CopyInfo::buildTensorString(srcBuffer, srcTensor);
+  copyInfo->dstStrKey = Profiler::CopyInfo::buildTensorString(dstBuffer, dstTensor);
+  m_total_copy_size += length;
+  if (m_profile_options & ProfileOptions::USE_INTERVALS) {
+    copyInfo->signpostId = beginSignpostInterval(SignpostTypes::BLIT_COPY, copyInfo->toString());
+  }
+  std::lock_guard<std::mutex> lock(m_mutex);
+  // this should not happen since we erase the copy info after profiling/logging it.
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(m_copy_info_list.count(dstBuffer) == 0);
+  m_copy_info_list.emplace(dstBuffer, std::move(copyInfo));
+}
+
+void MPSProfiler::endProfileCopy(const void* handle) {
+  Profiler::CopyInfo* copyInfo = nullptr;
+  {
+    std::lock_guard<std::mutex> lock(m_mutex);
+    TORCH_INTERNAL_ASSERT_DEBUG_ONLY(m_copy_info_list.count(handle), "Failed to get copy information!");
+    copyInfo = m_copy_info_list[handle].get();
+  }
+  addProfilerCompletedHandler(copyInfo, SignpostTypes::BLIT_COPY, SyncType::COMMIT_AND_WAIT);
+}
+
+void MPSProfiler::addProfilerCompletedHandler(Profiler::BaseInfo* info, SignpostTypes signpostType, SyncType syncType) {
+  const uint64_t signpostId = info->signpostId;
+  // signpost ID is used only for interval-based signposts, and must be non-zero
+  if (m_profile_options & ProfileOptions::USE_INTERVALS) {
+    TORCH_INTERNAL_ASSERT_DEBUG_ONLY(signpostId, "Signpost interval has no identifier!");
+  }
+  info->signpostId = 0; // reset signpostId for sanity check on next interval
+
+  auto m_stream = getDefaultMPSStream();
+  // NOTE: the following block isn't thread-safe, so make sure to capture profiler's info locally
```

Comment:
This looks correct to me. Sorry if I mis-understood yesterday. I thought you weren't using these instances

Reply:
Was the issue with gpuTime discrepancy discussed yesterday, solved ?

Reply:
We need to add scheduleHandlers to get accurate time for GPU of ops with signposts. It should resolve the tracing intervals issues when visualizing on tools.